### ISSUE TITLE
[Feature Request]:  Reduce JASP file size

### ISSUE THREAD

**Issue by Franck6S** [jasp-stats/jasp-issues#2808]
### Description

JASP File size

### Purpose

Usage & Storage of JASP files

### Use-case

all JASP files

### Is your feature request related to a problem?

My problem is the speed and usage of JASP, and related storage of information

### Is your feature request related to a JASP module?

Unrelated

### Describe the solution you would like

Long response calculation and size storage constraint on IT equipments

### Describe alternatives that you have considered

testing with other software compare to JASP

### Additional context

following image will show you related different file sizes with the same information :
![image](https://github.com/jasp-stats/jasp-issues/assets/152309774/ab1f2e1c-3160-4e34-a7d4-239b41063499)
You could see  x4 to 6 file size difference with JASP vs Excel

and close to double Minitab vs JASP : 
![image](https://github.com/jasp-stats/jasp-issues/assets/152309774/716a84b3-f24d-4b01-b9f8-81c359d1e30d)

**Comment by tomtomme** [comment]
The .jasp file you point out only contain data and no analysis or graphs?

**Comment by Franck6S** [comment]
They contain data and SPC chart (Quality control module), but I have always the same situation (big size file)  with other simple functions utilised : descriptives ANOVA, distributions.... here exemple with regression : data only in excel and JASP Linear regression : 
![image](https://github.com/jasp-stats/jasp-issues/assets/152309774/f19b97a2-d4d9-4728-aa34-8ac33dd70567)
FYI : Supporting JASP deployment, I get several "complaint" regarding JASP file sizes (storages sizes, sending files issues (limitation of size message...),...) which was not the case with previous software.

**Comment by tomtomme** [comment]
Currently you compare apples to oranges, so to say. A .jasp file can contain much more than the data like graphs, thus it will be bigger always. Sorry but there is nothing we can do about this. 

To have a fair comparison of file sizes you would need to add graphs of same quality and analysis like regression in excel too and store it as xlsx. 
 
If you compare a csv exported from jasp and a csv exported from excel the file size will be the same.
And no, a csv is not an excel file. It is an ancient file format that existed even long before excel was born.

**Comment by tomtomme** [comment]
Regarding minitab. Does a minitab file also only contain the data? If it also stores analysis and graphs, are they of the same quality?

**Comment by Franck6S** [comment]
I agree with excel datafile and JASP file analysis, but problem stay the same while compariing JASP and Minitab for the exact same content , and you could see that we are at close to doubling the size. Be aware that this is a real issue.

**Comment by Franck6S** [comment]
example from this week : issue found : error due to space of disk.
Main reason is when we generate several JASP Analysis (meaning  JASP Files) we rapidly use a big space and we get error message or not (last week : JASP roaming without interuption for several minutes)  : 
![image](https://github.com/jasp-stats/jasp-issues/assets/152309774/6216e23d-540a-4005-8642-7d7338868b90)

Let us know if you could consider the action in reducing JASP file size in the future.

**Comment by boutinb** [comment]
1577 GBytes is indeed huge!!
But the JASP files you listed above were ~30 MBytes, this is totally another kind of order. So there are here 2 different things here.
A JASP file is in fact a ZIP files containing the sqlite database, some description files, some png files if plots are used, and a state. What you can do is just to rename a JASP file with a zip suffix, and unzip it, then you can see its content, and check what takes most of the space (the json description files are zipped so they don't take so much space, but the database, png and state files are binaries).
For the error with the 1577 GBytes, this seems to be something else. Can you tell us exactly how big is your data files (how many columns and rows), and what do you do precisely in the ANOVA analysis to get this error?

**Comment by Franck6S** [comment]
It is a different case, the person having this error message was trying to run an Anova, and it was missing space on his computer, and you have seen related error message (1577 GB...); the related file that I compute was 
![image](https://github.com/jasp-stats/jasp-issues/assets/152309774/450ad672-5584-4912-a9bf-7e132a0b909c)
I suppose the error highlight by JASP was not relevant.
Globally, the main reason is that JASP files are taking bigger space that we used to do, and it generates after some time space disk issues. and then we get those errors. My request is to consider file size being smaller for the future JASP upgrades

**Comment by tomtomme** [comment]
Again, your observation is totally correct. But as @boutinb explained, the .JASP file format is already a compressed .ZIP and there is little we can do about its size. JASP provides many features with this file format, maybe more than MINITAB. Downsizing the file format would probably mean cutting features. But that cannot be the goal here, correct?

**Comment by EJWagenmakers** [comment]
I did notice that even for simple scenarios, JASP files can become relatively large. It would be good to study such cases by unzipping the file and assessing what components take up so much space. I suspect that the figures take up much space, but that can't be all. Some interesting detective work to be done here!

**Comment by EJWagenmakers** [comment]
...for instance, if an analysis uses MCMC or some other numerical method, it is probably the case that all of those samples are stored together with the output.

**Comment by shun2wang** [comment]
The main part of the jasp file is the `state` file, which stores R object-related things (such as plots) to speed up some processes. I didn't study it in depth. In fact, in most cases it seems that we can reduce writing some things to the `state` or merge similar content.

**Comment by Franck6S** [comment]
@Tomtomme, Agree that we do not want to remove feature even graphs (that we aim to develop more). The main purpose is just to make you aware about this situation (JASP file size) and looking in how this could be improved.

**Comment by RensDofferhoff** [comment]
I talked about this with Joris. I suggested getting rid of state files above 5MB.
This would mean you have a recompute when altering the options after loading a jasp file.
For some analysis stuff gets out of hand even for computationally light work.
We should discuss this for 0.20

**Comment by JorisGoosen** [comment]
> I talked about this with Joris. I suggested getting rid of state files above 5MB. This would mean you have a recompute when altering the options after loading a jasp file. For some analysis stuff gets out of hand even for computationally light work. We should discuss this for 0.20

Its all very analysis dependent btw, the worst offenders are the onces with lots of sampling (bootstrapping, mcms etc).
And then store that in the state (predictive analytics is the worst offender for that I think).

But the problem is also through the plots, they are stored in the state as well and due to the way ggplot works it can store lots of stuff from the r environment secretly. The way around that would be to call the plots only from sanitized R-function-scopes. @vandenman could weigh in on that.

But in any case, you will see that the filesize between different jasp-files varies wildly depending on the module+analysis used in them. It might be worthwile to investigate a bit and see which ones are the worst and focus some attention on that.

The solution of Rens would be very general though and is not a bad idea, but we probably want an option in the preferences to allow saving them anyway. So you dont have to wait so long when rerunning something big?
Or maybe ask the user during saving if they prefer a smaller file over saved time the next time they open it?
If the state file for instance turns out big?

**Comment by vandenman** [comment]
Having some option/ preference to purge the state is a good idea, although we should warn people accordingly (some people run analyses for days so we shouldn't make it to easy to throw that away). Reviewing the state management in the R analyses is also a good idea, but this may be a bit more tricky. In particular the plots tend to store more than is needed but this is hard to debug.

**Comment by shun2wang** [comment]
For the plots: we also stored .png image in the .jasp files. forgive my ignorance here😜, I'm not sure why we need to store so many plots objects inside state since I think it's fast enough to render many plots while refreshing analysis?

**Comment by tomtomme** [comment]
Bit what if you want to open/view an (old) jasp result without refreshing e.g. on OSF? Would then the images still be displayed? If not, this should also be made pretty clear to the user, what he/she is about to loose.

**Comment by JorisGoosen** [comment]
The pngs wont be removed in any case ;)
So OSF will still show the preview.

Just rerunning/resizing the plot would require rerunning the entire analysis.
If you want to experience this all you need to do is open a file from an older JASP-version and it will act as if the plot/state objects are missing and prompt you to rerun the analysis on a resize or something.

**Comment by JorisGoosen** [comment]
> Having some option/ preference to purge the state is a good idea, although we should warn people accordingly (some people run analyses for days so we shouldn't make it to easy to throw that away). 

Yeah, this is what I meant with "Or maybe ask the user during saving if they prefer a smaller file over saved time the next time they open it?"

So when saving a file we can check whether the state is huge or not, if huge (>1MB ?) and then ask the user whether they would like to store that or not. Mentioning that if they waited a long while for the analysis to run they probably want to.

We could even measure the runtime of an analysis and store the max runtime of one. (Since last refresh) Then if that is over 1 hour (or whatever) we can ask the user as well but being more specific.

> Reviewing the state management in the R analyses is also a good idea, but this may be a bit more tricky. In particular the plots tend to store more than is needed but this is hard to debug.

Maybe we should just stop storing the plot objects at all?

**Comment by shun2wang** [comment]
>Maybe we should just stop storing the plot objects at all?

I think so, then we just render .png statistic in result HTML, re-generate plots should be triggered by resizing plot or running analysis.

**Comment by JorisGoosen** [comment]
> I think so, then we just render .png statistic in result HTML

As I mentioned above, this is already the case.
The plot-object is only needed for resizing and plot-editing.

**Comment by JorisGoosen** [comment]
I mostly fixed some issues where the database was growing out of control (but that was caused by changes in the development version).
So this was never released and cant be the problem here.

We could surely add a checkbox to the workspace setting saying something like "Minimize filesize" where if enbled it won't store the plotobjects and stateobjects, just the results, the data and the analysissettings.
This will allow you to load the analysis fine. You would then just have to rerun everything from scratch on reload.

Which to be honest usually shouldnt really be too much of a problem. Except for some analyses where there is a long bootstrapping phase that is stored. (And which might by the way increase the filesize to ridiculous sizes, im looking at you predictive analytics)

**Comment by Franck6S** [comment]
Dears, FYI : I will have a look at it with the new version coming (will have some availability in coming weeks). Thanks your for effort.
